library/kotlin/src/io/envoyproxy/envoymobile/GRPCResponseHandler.kt

```diff
@@ -0,0 +1,177 @@
+package io.envoyproxy.envoymobile
+
+import io.envoyproxy.envoymobile.engine.types.EnvoyError
+import io.envoyproxy.envoymobile.engine.types.EnvoyErrorCode
+import java.io.ByteArrayOutputStream
+import java.nio.ByteBuffer
+import java.util.concurrent.Executor
+
+
+class GRPCResponseHandler(
+    val executor: Executor
+) {
+
+  /**
+   * Represents the process state of the response stream's body data.
+   */
+  private sealed class ProcessState {
+    // Awaiting a gRPC compression flag.
+    object CompressionFlag : ProcessState()
+
+    // Awaiting the length specification of the next message.
+    object MessageLength : ProcessState()
+
+    // Awaiting a message with the specified length.
+    class Message(val messageLength: Int) : ProcessState()
+  }
+
+  internal val underlyingHandler: ResponseHandler = ResponseHandler(executor)
+
+  private var errorClosure: (error: EnvoyError) -> Unit = { }
+  /**
+   * Specify a callback for when response headers are received by the stream.
+   * If `endStream` is `true`, the stream is complete.
+   *
+   * @param closure: Closure which will receive the headers, status code,
+   *                 and flag indicating if the stream is complete.
+   * @return ResponseHandler, this ResponseHandler.
```

Comment:
s/ResponseHandler/GRPCResponseHandler or just "response handler" (here and throughout the file)